### ISSUE TITLE
Solve a COMPLEX sparse system of equations using AMGCL's MPI. There is a data type mismatch when compiling, how can I solve it?

### ISSUE THREAD

**Issue by licongsheng** [ddemidov/amgcl#217]
When I use AMGCL to solve the complex sparse equation system, I wrote the following code with reference to poisson3Db_mpi.cpp,
```cpp
void CAMGCLSolver::Run(int argc, char *argv[])
{    // Show the name of the GPU we are using:


    amgcl::mpi::init mpi(&argc, &argv);
    amgcl::mpi::communicator world(MPI_COMM_WORLD);
	// The profiler:
    amgcl::profiler<> prof("MPI_SPFD");

    // Read the system matrix and the RHS:
    prof.tic("read");
    // Get the global size of the matrix:
    ptrdiff_t rows = matrixM;
    ptrdiff_t cols = 1;

    // Split the matrix into approximately equal chunks of rows
    std::cout<<"Size of world is " << world.size << ", Rows of system matrix is " << rows << std::endl;
    ptrdiff_t chunk = (ptrdiff_t)((float)rows / (float)world.size + 0.5);
    ptrdiff_t row_beg = std::min(rows, chunk * world.rank);
    ptrdiff_t row_end = std::min(rows, row_beg + chunk);
    chunk = row_end - row_beg;
    std::cout<<"Rank " << world.rank << " From " << row_beg << " to " << row_end << "..." << std::endl;

    // Read our part of the system matrix and the RHS.
    std::vector<ptrdiff_t> ptr, col;
    std::vector< complex<double> > val, rhs;
    //Sperate total system to every cluster
    std::cout<<"-------------------------------------------------------------" << std::endl;
    std::cout<<"parse crs matrix ... " << std::endl;
    // parse_crs() and parse_dense()  are rewritten from read_crs() and read_dense()
    amgcl::parse_crs(matrixM, coo_Row, csr_Col, matrix_A, rows, ptr, col, val, row_beg, row_end, world.rank);
    std::cout<<"parse dense vector ... " << std::endl;
    amgcl::parse_dense(matrixM, 1, b, rows, cols, rhs, row_beg, row_end, world.rank);
    prof.toc("read");

    if (world.rank == 0)
            std::cout
                << "World size: " << world.size << std::endl
                << "Matrix " << ": " << rows << "x" << rows << std::endl
                << "RHS " << ": " << rows << "x" << cols << std::endl;

	// Compose the solver type
	// the preconditioner backend:
	typedef amgcl::backend::builtin<complex<double>> PBackend;

	std::cout << "World Rand " << world.rank << " make solver" << std::endl;
	typedef amgcl::mpi::make_solver<
	        amgcl::mpi::amg<
	        		PBackend,
	            amgcl::mpi::coarsening::smoothed_aggregation<PBackend>,
	            amgcl::mpi::relaxation::ilu0<PBackend>   //ilu0 spai0
	            >,
	        amgcl::mpi::solver::idrs<PBackend>   // bicgstab  gmres idrs
	        > Solver;

	 //Solver parameters
	 Solver::params prm;
	 prm.solver.maxiter=1000;
	 prm.solver.verbose=true;

	 std::cout<<"-------------------------------------------------------------" << std::endl;
	 std::cout << "World Rand " << world.rank << " Create the distributed matrix from the local parts" << std::endl;
	 // Create the distributed matrix from the local parts.
	 std::cout << "World Rand " << world.rank << " Chunk: " << chunk << std::endl <<
			 "                  rows from PRT: " << ptr.size() << std::endl <<
			 "                  rows from col: " << col.back() << std::endl <<
			 "                  nnz Size from PRT: " << ptr.back() << std::endl <<
			 "                  nnz Size from col: " << col.size() << std::endl <<
			 "                  nnz Size from val: " << val.size() << std::endl;

	std::cout<<"-------------------------------------------------------------" << std::endl;
	auto A = std::make_shared<amgcl::mpi::distributed_matrix<PBackend>>(
			 	 world, std::tie(chunk, ptr, col, val));

	std::cout << "World Rand " << world.rank << " Partition the matrix and the RHS vector" << std::endl;
	// Partition the matrix and the RHS vector.
	// If neither ParMETIS not PT-SCOTCH are not available,
	// just keep the current naive partitioning.
	#if defined(AMGCL_HAVE_PARMETIS) || defined(AMGCL_HAVE_SCOTCH)
		#  if defined(AMGCL_HAVE_PARMETIS)
			typedef amgcl::mpi::partition::parmetis<SBackend> Partition;
		#  elif defined(AMGCL_HAVE_SCOTCH)
			typedef amgcl::mpi::partition::ptscotch<SBackend> Partition;
		#  endif

		if (world.size > 1) {
			prof.tic("partition");
			Partition part;

		   // part(A) returns the distributed permutation matrix:
		   auto P = part(*A);
		   auto R = transpose(*P);

		   // Reorder the matrix:
		   A = product(*R, *product(*A, *P));

		   // and the RHS vector:
		   std::vector<complex<double>> new_rhs(R->loc_rows());
		   R->move_to_backend(typename PBackend::params());

//		   complex<double> one(1.0, 0.0);
//		   complex<double> zero(0.0, 0.0);
//		   amgcl::backend::spmv(one, *R, rhs, zero, new_rhs);
	      amgcl::backend::spmv(1, *R, rhs, 0, new_rhs);
		   rhs.swap(new_rhs);

		   // Update the number of the local rows
		   // (it may have changed as a result of permutation):
		   chunk = A->loc_rows();
		   prof.toc("partition");
			}
	#endif

	std::cout << "World Rand " << world.rank << " Initialize the solver" << std::endl;
	// Initialize the solver:
	prof.tic("setup");
	//Solver solve(world, A, prm);
	Solver solve (world, A);
	prof.toc("setup");

	// Show the mini-report on the constructed solver:
	if (world.rank == 0)
		std::cout << solve << std::endl;

	std::cout << "World Rand " << world.rank << " Solve the system with the zero initial approximation" << std::endl;
	// Solve the system with the zero initial approximation:
	int iters;
	double error;
	std::vector<complex<double>> x(chunk, 0.0);

	prof.tic("solve");
	//< PBackend, std::vector<complex<double>>, std::vector<complex<double>> >
	std::tie(iters, error) = solve(*A, rhs, x);
	prof.toc("solve");
	for (int i = row_beg; i < row_end; i++)
		h_x[i] = x[i];
	// Output the number of iterations, the relative error,
	// and the profiling data:

	if (world.rank == 0)
	{
		for (int i = 0; i < matrixM; i++)
			h_x[i] = x[i];
		delete[]real_h_x; real_h_x = NULL;
	    std::cout << "Iters: " << iters << std::endl
	    			<< "Error: " << error << std::endl
	             << prof << std::endl;

	}
	//delete[]h_b; h_b = NULL;

}
```
but the following error occurs when compiling, how can I solve it? 
```
../src/CAMGCLSolver.cpp:373:43:   required from here
/usr/local/amgcl/include/amgcl/backend/detail/_**matrix_ops.hpp:71:30: error: no match for ‘operator*’ (operand types are ‘int’ and ‘V {aka std::complex<double>}’)
                 y[i] = alpha * sum + beta * y[i];
                        ~~~~~~^~~~~**_
In file included from /home/SPFD_GMGA/mpi_QSMF_Solver/src/CAMGCLSolver.h:11:0,
                 from ../src/CAMGCLSolver.cpp:8:
/opt/rh/devtoolset-7/root/usr/include/c++/7/complex:386:5: note: candidate: template<class _Tp> std::complex<_Tp> std::operator*(const std::complex<_Tp>&, const std::complex<_Tp>&)
     operator*(const complex<_Tp>& __x, const complex<_Tp>& __y)
     ^~~~~~~~
/opt/rh/devtoolset-7/root/usr/include/c++/7/complex:386:5: note:   template argument deduction/substitution failed:
In file included from /usr/local/amgcl/include/amgcl/backend/builtin.hpp:50:0,
                 from /usr/local/amgcl/include/amgcl/value_type/complex.hpp:34,
                 from ../src/CAMGCLSolver.cpp:12:
/usr/local/amgcl/include/amgcl/backend/detail/matrix_ops.hpp:71:30: note:   mismatched types ‘const std::complex<_Tp>’ and ‘int’
                 y[i] = alpha * sum + beta * y[i];
```
Thanks.

**Comment by ddemidov** [comment]
Please either provide a minimal reproducible code that I could try to compile or the complete error log.

**Comment by ddemidov** [comment]
Could you check if 1d15178a8a6b569ceb06f88ebac56f6d3f6bb38b helps?

**Comment by licongsheng** [comment]
The following is my simplified code, the compilation can reproduce the error record, please help to check. 

[main.txt](https://github.com/ddemidov/amgcl/files/8013192/main.txt)

In addition, I am using the latest AMGCL. I checked the content in the relevant files. The relevant content of [1d15178](https://github.com/ddemidov/amgcl/commit/1d15178a8a6b569ceb06f88ebac56f6d3f6bb38b) is the latest code in the idr.hpp file in AMGCL.

**Comment by ddemidov** [comment]
Thanks, I can not reproduce the compilation error with your file and 1d15178a8a6b569ceb06f88ebac56f6d3f6bb38b (I do see the error without the commit).

**Comment by licongsheng** [comment]
In addition, when I did not use MPI, there was no problem, and the result was correct.

**Comment by ddemidov** [comment]
The latest commit in amgcl (1d15178a8a6b569ceb06f88ebac56f6d3f6bb38b) should fix your issue. I can compile your code without errors.

**Comment by licongsheng** [comment]
Thank you very much for your guidance. There is no problem with the compilation at present. The execution of the program is interrupted. I will debug it again.
![mpi_error](https://user-images.githubusercontent.com/26063060/152891815-e92df58e-a7d6-45d3-b711-96194d96deb0.png)

**Comment by licongsheng** [comment]
Thanks to ddemidov for the AMGCL calculation library, I basically achieved what I needed to do. In order to facilitate other developers who may have similar problems in their applications, when AMGCL supports complex data types, it is necessary to make the following additions to <amgcl/mpi/util.hpp>.

![util修改](https://user-images.githubusercontent.com/26063060/152901540-10876180-21cc-4a5e-a8b7-3a9faba3b437.png)

In addition, the io for AMGCL only provides the read matrix from the file. I share the code of the direct memory vector<> for MPI data allocation as follows, so that other developers can directly develop MPI applications.

------------------------------------------------------------------------------------------------------------------------
[parse_crs_matrix_mpi.txt](https://github.com/ddemidov/amgcl/files/8019968/parse_crs_matrix_mpi.txt)
--------------------------------------------------------------------------------------------------------------------------

**Comment by ddemidov** [comment]
Thank you, I have added mpi support for complex datatypes in 9c3cfca969997a939b6c504502648e599b162d5e.

**Comment by ddemidov** [comment]
Looks like you are the first one to try MPI solver with a complex system :).

Sorry, I did not try to compile your example, but I have introduced a complex-valued MPI example in a95dae7fa4907bb6da9158656c641e3138a7cf8d, and I think I hit the same bug you did. It should be fixed in 75b52cea4d35df42b4c6ce5343537159b1f3f3ce. Could you please check your system with the latest amgcl version?
